### ISSUE TITLE
Apply the get_most_stable_tautomer function to pandas dataframe (from csv)

### ISSUE THREAD

**Issue by mmagithub** [isayevlab/Auto3D_pkg#71]
Hi,

I am wondering if there is a straightforward way to use the get_most_stable_tautomer function as a standalone rdkit function that can be applied to a SMILES column in a pandas dataframe, for example:

df['most_stable_taut'] = df['smiles'].apply(get_most_stable_tautomer)

It seems the current use of the function requires a strict input file format (smiles / sdf), and i cannot find a typical use case to satisfy what I am looking for.

Any clue ?

Thanks,
M

**Comment by LiuCMU** [comment]
I guess you meant `get_stable_tautomers`. The initial design takes a smi/sdf file as the input file because auto3d can process them in parallel. This will be much more efficient than processing a molecule at a time.

It is indeed a good idea to make it applicable on a pandas dataframe. The only concern is whether the function is fast enough. For each tautomer, auto3d needs to search, optimize and compare conformers, which is a relative expensive process to be broadcasted to individual molecules.

**Comment by mmagithub** [comment]
Thanks for the answer, this repo is amazing.
Considering pandas' powerful capabilities for handling and manipulating large datasets, including features that support parallel processing, I believe a tailored solution might be within reach. Specifically, a wrapper function for the (get_stable_tautomers) code designed to work directly on the SMILES column of a pandas DataFrame could significantly streamline the process. This function would essentially bridge the gap between pandas' data management and auto3d's parallel processing capabilities. I am wondering if you have a code snippet to do that.

**Comment by LiuCMU** [comment]
Thanks for the encouragement!

When applying a function to a specific column of a data frame, for example `df[col].apply(func)`, the function `func` itself takes a single element of the column as the input. Data frame is able to broadcast the function to all elements in the column efficiently. So, the function itself does not have a parallel feature, but the data frame is able to run the function efficiently through some special data management. 

If `get_stable_tautomers` is to be applied to a column of a data frame, then the input should be a single SMILES. However, `get_stable_tautomers` is more efficient when processing a batch of SMILES together so that it can optimize all conformers at the same time. That's why this function takes a smi file of molecules.

In short, using data frame's data management would disable the parallel capacity of auto3d. I'm not sure if there is a way to work around this. If so, I'd love to learn about it :-)

**Comment by isayev** [comment]
Great question! Here's how to apply tautomer enumeration to a pandas DataFrame:

## Option 1: Using RDKit directly (recommended for DataFrames)

```python
import pandas as pd
from rdkit import Chem
from rdkit.Chem.MolStandardize import rdMolStandardize

def get_canonical_tautomer(smiles):
    """Get the canonical (most stable) tautomer for a SMILES string."""
    try:
        mol = Chem.MolFromSmiles(smiles)
        if mol is None:
            return None
        enumerator = rdMolStandardize.TautomerEnumerator()
        canonical = enumerator.Canonicalize(mol)
        return Chem.MolToSmiles(canonical)
    except Exception:
        return None

# Apply to DataFrame
df = pd.DataFrame({'smiles': ['CC(=O)Nc1ccc(O)cc1', 'c1ccccc1O', 'CC(=O)C']})
df['canonical_tautomer'] = df['smiles'].apply(get_canonical_tautomer)
print(df)
```

## Option 2: Using Auto3D for energy-ranked tautomers (batch processing)

If you need **energy-ranked** tautomers (not just canonical), write to a temp file and use Auto3D:

```python
import pandas as pd
import tempfile
import os
from Auto3D.auto3D import options, main
from Auto3D.tautomer import get_stable_tautomers
from rdkit import Chem

def process_batch_tautomers(df, smiles_col='smiles', id_col=None):
    """Process a DataFrame batch through Auto3D tautomer ranking."""
    with tempfile.TemporaryDirectory() as tmpdir:
        # Write SMILES to temp file
        input_path = os.path.join(tmpdir, 'input.smi')
        with open(input_path, 'w') as f:
            for idx, row in df.iterrows():
                name = row[id_col] if id_col else f'mol_{idx}'
                f.write(f"{row[smiles_col]} {name}\n")
        
        # Run Auto3D tautomer ranking
        args = options(input_path, k=1, enumerate_tautomer=True, 
                      tauto_engine='rdkit', use_gpu=False)
        result_sdf = get_stable_tautomers(args, tauto_k=1)
        
        # Read results back
        results = {}
        for mol in Chem.SDMolSupplier(result_sdf, removeHs=False):
            if mol:
                name = mol.GetProp('_Name')
                results[name] = Chem.MolToSmiles(Chem.RemoveHs(mol))
        
        return results

# Usage
df = pd.DataFrame({'smiles': ['CC(=O)Nc1ccc(O)cc1'], 'id': ['paracetamol']})
tautomers = process_batch_tautomers(df, smiles_col='smiles', id_col='id')
```

**Note:** For large DataFrames, Option 1 (RDKit canonical tautomer) is much faster. Use Option 2 only when you need energy-based ranking.

Closing as answered!
